third/nmh/man/show.man
.\"
.\" %nmhwarning%
.\" $Id: show.man,v 1.1.1.1 1999-02-07 18:14:22 danw Exp $
.\"
.\" include the -mh macro file
.so %etcdir%/tmac.h
.\"
.TH SHOW %manext1% MH.6.8 [%nmhversion%]
.SH NAME
show \- show (display) messages
.SH SYNOPSIS
.in +.5i
.ti -.5i
show
\%[+folder] \%[msgs]
\%[\-showproc\ program]
.br
\%[\-showmimeproc\ program]
\%[\-header] \%[\-noheader]
.br
\%[\-draft]
\%[\-checkmime] \%[\-nocheckmime]
.br
\%[switches\ for\ \fIshowproc\fR or \fIshowmimeproc\fR]
.br
\%[\-version]
\%[\-help]
.in -.5i
.SH DESCRIPTION
\fIShow\fR lists each of the specified messages to the standard output
(typically, the terminal).

By default, text (non-MIME) messages are filtered and displayed by
the \fInmh\fR command \fImhl\fR.  This command will display text
messages in a nice, uniform format.  It also allows you to configure
the format of the displayed messages and which headers fields are
shown.  See the \fImhl\fR(1) manual page for the details about this
command.  This default can be changed by defining the \fIshowproc\fR
profile component.  Any switches not recognized by \fIshow\fR are
passed along to that program.  To override the default and the
\fIshowproc\fR profile component, use the `\-showproc\ program'
switch.  For example, `\-showproc\ more' will cause the \fImore\fR(1)
program to list the messages with no reformatting.  Normally, this
program is specified as the \fIshowproc\fR in the user's
\&.mh\(ruprofile, rather than using a command line switch.

By default, non-text messages (MIME messages with multi-media
contents) are processed and displayed by the \fInmh\fR command
\fImhshow\fR.  See the \fImhshow\fR(1) manual page for details
about this command.  This default can changed by defining the
\fIshowmimeproc\fR profile component.  Any switches not recognized
by \fIshow\fR are passed along to that program.  To override this
default and the \fIshowmimeproc\fR profile component, use the
`\-showmimeproc\ program' switch.

Note that in some cases, \fIshow\fR may invoke the \fIshowmimeproc\fR
even for textual contents.  This will happen for text messages that
specify a transfer encoding (such as MIME quoted-printable or
base64) or specify a character set that \fIshow\fR doesn't believe
can be displayed natively.  The environment variable MM_CHARSET
should be set to the terminal's native character set to avoid
gratuitous invocations of the \fIshowmimeproc\fR.  See the
mh-profile(5) man page for details about this environment variable.

The option `\-checkmime' (set by default) instructs \fIshow\fR to
test if any of the messages to be displayed are non-text (MIME)
messages.  If any are non-text, they are displayed by the program
\fIshowmimeproc\fR, else they are displayed by the program
\fIshowproc\fR.  The option `-nocheckmime' disables this test and
instructs \fIshow\fR to use \fIshowproc\fR, regardless of whether
any of the messages are non-text (MIME) messages.

The `\-noshowproc' switch will disable any formatting or paging of
messages.  It is equivalent to `-nocheckmime\ -showproc\ cat'.  It
is still accepted, but should be considered (somewhat) obsolete.

If the environment variable \fBNOMHNPROC\fR is set, the test for
non-text (MIME) messages will be disabled.  This method is obsolete.
Use the `-nocheckmime' switch instead.

The `\-header' switch tells \fIshow\fR to display a one\-line
description of the message being shown.  This description includes
the folder and the message number.

If no `msgs' are specified, the current message is used.  Although
it depends on the specific \fIshowproc\fR or \fIshowmimeproc\fR,
in the default setup when more than one message is specified, you
will be prompted for a <RETURN> prior to listing each message.
Each message will be listed a page at a time, and when the end of
page is reached, the program will wait for a <SPACE> or <RETURN>.
If a <RETURN> is entered, it will print the next line, whereas
<SPACE> will print the next screenful.

If the standard output is not a terminal, no queries are made, and
each file is listed with a one\-line header and two lines of
separation.

\*(lqshow \-draft\*(rq will list the file <mh\-dir>/draft if it
exists.

If the profile entry \*(lqUnseen\-Sequence\*(rq is present and
non\-empty, then \fIshow\fR will remove each of the messages shown
from each sequence named by the profile entry.
.Fi
^$HOME/\&.mh\(ruprofile~^The user profile
.Pr
^Path:~^To determine the user's nmh directory
.Ps
^Current\-Folder:~^To find the default current folder
.Ps
^Unseen\-Sequence:~^To name sequences denoting unseen messages
.Ps
^showproc:~^Program to show text (non-MIME) messages
.Ps
^showmimeproc:~^Program to show non-text (MIME) messages
.Sa
mhl(1), mhshow(1), more(1), next(1), prev(1), scan(1)
.De
`+folder' defaults to the current folder
.Ds
`msgs' defaults to cur
.Ds
`\-checkmime'
.Ds
`\-header'
.Co
If a folder is given, it will become the current folder.  The last
message shown will become the current message.
.Bu
The `\-header' switch doesn't work when `msgs' expands to more than
one message.  If the \fIshowproc\fR is \fImhl\fR, then is problem can
be circumvented by referencing the \*(lqmessagename\*(rq field in the
\fImhl\fR format file.

\fIShow\fR updates the user's context before showing the message.
Hence \fIshow\fR will mark messages as seen prior to the user actually
seeing them.  This is generally not a problem, unless the user relies
on the \*(lqunseen\*(rq messages mechanism, and interrupts \fIshow\fR
while it is showing \*(lqunseen\*(rq messages.

If your \fIshowproc\fR is \fImhl\fR (the default), then \fIshow\fR uses
a built\-in \fImhl\fR: it does not actually run the \fImhl\fR program.
Hence, if you define your own \fIshowproc\fR, don't call it \fImhl\fR
since \fIshow\fR won't run it.

If your \fIshowproc\fR is the pager \fImore\fR, then avoid running
\fIshow\fR in the background with only its standard output piped to
another process, as in

.ti +.5i
show | imprint &

Due to a bug in \fImore\fR, show will go into a \*(lqtty input\*(rq state.
To avoid this problem, re\-direct \fIshow\fR's diagnostic output as well.
For users of \fIcsh\fR:

.ti +.5i
show |& imprint &

For users of \fIsh\fR:

.ti +.5i
show 2>&1 | imprint &
.En


doc/news
This file is intended to keep track of user-visible changes to the
release, for the construction of the Athena release notes.

Changes from release 8.0 to release 8.1:

	* Athena version numbers are now dotted triplets, e.g. 8.1.0
	  or 8.1.12.  Paths which include all or part of the Athena
	  version number now take the form "pathname-version",
	  e.g. "srvd-8.1" or "dash.menus-8.1"

	* In release 8.0, running "xmkmf" with the default path would
	  generate a Makefile which looked for its include files and
	  libraries in /mit/x11, which could be one of several
	  lockers.  Now it is preferrable to use the vendor X
	  libraries, so this bit of glue is going away.  There will be
	  no /usr/athena/bin/xmkmf any more, nor any symlinks in
	  /usr/athena pointing to the x11 locker.

	  Users can still do something like "add -f x11r5" and then
	  "xmkmf", and things should work, if the x11r5 locker is
	  built for their platform.

	* Athena bitmaps have moved to /usr/athena/include/X11/bitmaps.
	  Previously, they were in /usr/athena/lib/X11/bitmaps before
	  because of the /mit/x11 glue.

	* The symlink /bin/csh -> /bin/athena/tcsh has gone away.  The
	  update to 8.0 will fix /bin/csh references in /etc/passwd
	  and /etc/passwd.local to refer to /bin/athena/tcsh instead.

	* emacs18 has gone away.

	* AFS includes and libraries are installed in
	  /usr/afsws/include and /usr/afsws/lib.

	* wcl includes and libraries are installed in /usr/athena.

	* The basic krb5 headers and libraries are installed in
	  /usr/athena.

	* emacs has been upgraded from 19.30 to 19.34.  There should
	  be no significant user-visible changes.

	* RCS has been upgraded from version 4 to version 5.7.

	* x3270 has been upgraded from version 1.2 to version 3.1.1.3.

	* sysinfo is now available under Irix.

	* /bin/athena/tcsh now has Hesiod support enabled, which means
	  that "~lockername" will work in a pathname in most cases.
	  (It won't work for users with local passwd information who
	  don't have their lockers attached.)

	* For improved ease of debugging, many binaries are installed
	  with symbol tables, and compiled without optimization.
	  Therefore, the contents of /bin/athena and /usr/athena/bin
	  may be somewhat larger, and the binaries may consume more
	  CPU cycles, although it's unlikely that anyone will notice
	  the speed difference.


third/nmh/uip/show.c

/*
 * show.c -- show/list messages
 *
 * $Id: show.c,v 1.3 1999-10-08 22:14:32 danw Exp $
 */

#include <h/mh.h>
#include <h/mime.h>

static struct swit switches[] = {
#define CHECKMIMESW          0
    { "checkmime", 0 },
#define NOCHECKMIMESW        1
    { "nocheckmime", 0 },
#define	HEADSW               2
    { "header", 0 },
#define	NHEADSW              3
    { "noheader", 0 },
#define	FORMSW               4
    { "form formfile", 0 },
#define	PROGSW               5
    { "moreproc program", 0 },
#define	NPROGSW              6
    { "nomoreproc", 0 },
#define	LENSW                7
    { "length lines", 0 },
#define	WIDTHSW              8
    { "width columns", 0 },
#define	SHOWSW               9
    { "showproc program", 0 },
#define SHOWMIMESW          10
    { "showmimeproc program", 0 },
#define	NSHOWSW             11
    { "noshowproc", 0 },
#define	DRFTSW              12
    { "draft", 0 },
#define	FILESW              13
    { "file file", -4 },		/* interface from showfile */
#define VERSIONSW           14
    { "version", 0 },
#define	HELPSW              15
    { "help", 0 },
    { NULL, 0 }
};

/*
 * static prototypes
 */
static int is_nontext(char *);

#define	SHOW  0
#define	NEXT  1
#define	PREV  2


int
main (int argc, char **argv)
{
    int draftsw = 0, headersw = 1, msgp = 0;
    int nshow = 0, checkmime = 1, mime;
    int vecp = 1, procp = 1, isdf = 0, mode = SHOW, msgnum;
    char *cp, *maildir, *file = NULL, *folder = NULL, *proc;
    char buf[BUFSIZ], **argp, **arguments;
    char *msgs[MAXARGS], *vec[MAXARGS];
    struct msgs *mp;

#ifdef LOCALE
    setlocale(LC_ALL, "");
#endif
    invo_name = r1bindex (argv[0], '/');

    /* read user profile/context */
    context_read();

    if (!strcasecmp (invo_name, "next")) {
	mode = NEXT;
    } else if (!strcasecmp (invo_name, "prev")) {
	mode = PREV;
    }
    arguments = getarguments (invo_name, argc, argv, 1);
    argp = arguments;

    while ((cp = *argp++)) {
	if (*cp == '-') {
	    switch (smatch (++cp, switches)) {
		case AMBIGSW: 
		    ambigsw (cp, switches);
		    done (1);
		case UNKWNSW: 
		case NPROGSW:
		    vec[vecp++] = --cp;
		    continue;

		case HELPSW: 
		    snprintf (buf, sizeof(buf),
			"%s [+folder] %s[switches] [switches for showproc]",
			invo_name, mode == SHOW ? "[msgs] ": "");
		    print_help (buf, switches, 1);
		    done (1);
		case VERSIONSW:
		    print_version(invo_name);
		    done (1);

		case DRFTSW: 
		    if (file)
			adios (NULL, "only one file at a time!");
		    draftsw++;
		    if (mode == SHOW)
			continue;
usage:
		    adios (NULL,
			    "usage: %s [+folder] [switches] [switches for showproc]",
			    invo_name);
		case FILESW: 
		    if (mode != SHOW)
			goto usage;
		    if (draftsw || file)
			adios (NULL, "only one file at a time!");
		    if (!(cp = *argp++) || *cp == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    file = path (cp, TFILE);
		    continue;

		case HEADSW: 
		    headersw++;
		    continue;
		case NHEADSW: 
		    headersw = 0;
		    continue;

		case FORMSW:
		    vec[vecp++] = --cp;
		    if (!(cp = *argp++) || *cp == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    vec[vecp++] = getcpy (etcpath(cp));
		    continue;

		case PROGSW:
		case LENSW:
		case WIDTHSW:
		    vec[vecp++] = --cp;
		    if (!(cp = *argp++) || *cp == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    vec[vecp++] = cp;
		    continue;

		case SHOWSW: 
		    if (!(showproc = *argp++) || *showproc == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    nshow = 0;
		    continue;
		case NSHOWSW: 
		    nshow++;
		    continue;

		case SHOWMIMESW:
		    if (!(showmimeproc = *argp++) || *showmimeproc == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    nshow = 0;
		    continue;
		case CHECKMIMESW:
		    checkmime++;
		    continue;
		case NOCHECKMIMESW:
		    checkmime = 0;
		    continue;
	    }
	}
	if (*cp == '+' || *cp == '@') {
	    if (folder)
		adios (NULL, "only one folder at a time!");
	    else
		folder = path (cp + 1, *cp == '+' ? TFOLDER : TSUBCWF);
	} else {
	    if (mode != SHOW)
		goto usage;
	    else
		msgs[msgp++] = cp;
	}
    }
    procp = vecp;

    /* If showing multiple messages, default to -nocheckmime.
     * (But just decrement it rather than setting to 0, so user
     * can specify -checkmime to override.
     */
    if (msgp > 1)
      checkmime--;

    if (!context_find ("path"))
	free (path ("./", TFOLDER));

    if (draftsw || file) {
	if (msgp)
	    adios (NULL, "only one file at a time!");
	vec[vecp++] = draftsw
	    ? getcpy (m_draft (folder, msgp ? msgs[0] : NULL, 1, &isdf))
	    : file;
	goto go_to_it;
    }

#ifdef WHATNOW
    if (!msgp && !folder && mode == SHOW && (cp = getenv ("mhdraft")) && *cp) {
	draftsw++;
	vec[vecp++] = cp;
	goto go_to_it;
    }
#endif /* WHATNOW */

    if (!msgp) {
	switch (mode) {
	    case NEXT:
		msgs[msgp++] = "next";
		break;
	    case PREV:
		msgs[msgp++] = "prev";
		break;
	    default:
		msgs[msgp++] = "cur";
		break;
	}
    }

    if (!folder)
	folder = getfolder (1);
    maildir = m_maildir (folder);

    if (chdir (maildir) == NOTOK)
	adios (maildir, "unable to change directory to");

    /* read folder and create message structure */
    if (!(mp = folder_read (folder)))
	adios (NULL, "unable to read folder %s", folder);

    /* check for empty folder */
    if (mp->nummsg == 0)
	adios (NULL, "no messages in %s", folder);

    /* parse all the message ranges/sequences and set SELECTED */
    for (msgnum = 0; msgnum < msgp; msgnum++)
	if (!m_convert (mp, msgs[msgnum]))
	    done (1);

    /*
     * Set the SELECT_UNSEEN bit for all the SELECTED messages,
     * since we will use that as a tag to know which messages
     * to remove from the "unseen" sequence.
     */
    for (msgnum = mp->lowsel; msgnum <= mp->hghsel; msgnum++)
	if (is_selected(mp, msgnum))
	    set_unseen (mp, msgnum);

    seq_setprev (mp);		/* set the Previous-Sequence */
    seq_setunseen (mp, 1);	/* unset the Unseen-Sequence */

    if (mp->numsel > MAXARGS - 2)
	adios (NULL, "more than %d messages for show exec", MAXARGS - 2);

    for (msgnum = mp->lowsel; msgnum <= mp->hghsel; msgnum++)
	if (is_selected(mp, msgnum))
	    vec[vecp++] = getcpy (m_name (msgnum));

    seq_setcur (mp, mp->hghsel);	/* update current message  */
    seq_save (mp);			/* synchronize sequences   */
    context_replace (pfolder, folder);	/* update current folder   */
    context_save ();			/* save the context file   */

    if (headersw && vecp == 2)
	printf ("(Message %s:%s)\n", folder, vec[1]);

go_to_it: ;
    fflush (stdout);

    vec[vecp] = NULL;

    /*
     * Decide which "proc" to use
     */
    mime = 0;
    if (nshow) {
	proc = catproc;
    } else {
	/* check if any messages are non-text MIME messages */
	if (checkmime && !getenv ("NOMHNPROC")) {
	    if (!draftsw && !file) {
		/* loop through selected messages and check for MIME */
		for (msgnum = mp->lowsel; msgnum <= mp->hghsel; msgnum++)
		    if (is_selected (mp, msgnum) && is_nontext (m_name (msgnum))) {
			mime = 1;
			break;
		    }
	    } else {
		/* check the file or draft for MIME */
		if (is_nontext (vec[vecp - 1]))
		    mime = 1;
	    }
	}

	/* Set the "proc" */
	if (mime)
	    proc = showmimeproc;
	else
	    proc = showproc;
    }

    if (folder && !draftsw && !file)
	m_putenv ("mhfolder", folder);

    /*
     * For backward compatibility, if the "proc" is mhn,
     * then add "-show" option.  Add "-file" if showing
     * file or draft.
     */
    if (strcmp (r1bindex (proc, '/'), "mhn") == 0) {
	if (draftsw || file) {
	    vec[vecp] = vec[vecp - 1];
	    vec[vecp - 1] = "-file";
	    vecp++;
	}
	vec[vecp++] = "-show";
	vec[vecp] = NULL;
    }

    /* If the "proc" is "mhshow", add "-file" if showing file or draft.
     */
    if (strcmp (r1bindex (proc, '/'), "mhshow") == 0 && (draftsw || file) ) {
       vec[vecp] = vec[vecp - 1];
       vec[vecp - 1] = "-file";
       vec[++vecp] = NULL;
    }

    /*
     * If "proc" is mhl, then run it internally
     * rather than exec'ing it.
     */
    if (strcmp (r1bindex (proc, '/'), "mhl") == 0) {
	vec[0] = "mhl";
	mhl (vecp, vec);
	done (0);
    }

    /*
     * If you are not using a nmh command as your "proc", then
     * add the path to the message names.  Currently, we are just
     * checking for mhn here, since we've already taken care of mhl.
     */
    if (!strcmp (r1bindex (proc, '/'), "mhl")
	    && !draftsw
	    && !file
	    && chdir (maildir = concat (m_maildir (""), "/", NULL)) != NOTOK) {
	mp->foldpath = concat (mp->foldpath, "/", NULL);
	cp = ssequal (maildir, mp->foldpath)
	    ? mp->foldpath + strlen (maildir)
	    : mp->foldpath;
	for (msgnum = procp; msgnum < vecp; msgnum++)
	    vec[msgnum] = concat (cp, vec[msgnum], NULL);
    }

    vec[0] = r1bindex (proc, '/');
    execvp (proc, vec);
    adios (proc, "unable to exec");
}

/*
 * Cheat:  we are loaded with adrparse, which wants a routine called
 * OfficialName().  We call adrparse:getm() with the correct arguments
 * to prevent OfficialName() from being called.  Hence, the following
 * is to keep the loader happy.
 */

char *
OfficialName (char *name)
{
    return name;
}


/*
 * Check if a message or file contains any non-text parts
 */
static int
is_nontext (char *msgnam)
{
    int	result, state;
    char *bp, *cp, *dp;
    char buf[BUFSIZ], name[NAMESZ];
    FILE *fp;

    if ((fp = fopen (msgnam, "r")) == NULL)
	return 0;

    for (state = FLD;;) {
	switch (state = m_getfld (state, name, buf, sizeof(buf), fp)) {
	case FLD:
	case FLDPLUS:
	case FLDEOF:
	    /*
	     * Check Content-Type field
	     */
	    if (!strcasecmp (name, TYPE_FIELD)) {
		int passno;
		char c;

		cp = add (buf, NULL);
		while (state == FLDPLUS) {
		    state = m_getfld (state, name, buf, sizeof(buf), fp);
		    cp = add (buf, cp);
		}
		bp = cp;
		passno = 1;

again:
		for (; isspace (*bp); bp++)
		    continue;
		if (*bp == '(') {
		    int i;

		    for (bp++, i = 0;;) {
			switch (*bp++) {
			case '\0':
invalid:
			    result = 0;
			    goto out;
			case '\\':
			    if (*bp++ == '\0')
				goto invalid;
			    continue;
			case '(':
			    i++;
			    /* and fall... */
			default:
			    continue;
			case ')':
			    if (--i < 0)
				break;
			    continue;
			}
			break;
		    }
		}
		if (passno == 2) {
		    if (*bp != '/')
			goto invalid;
		    bp++;
		    passno = 3;
		    goto again;
		}
		for (dp = bp; istoken (*dp); dp++)
		    continue;
		c = *dp;
		*dp = '\0';
		if (!*bp)
		    goto invalid;
		if (passno > 1) {
		    if ((result = (strcasecmp (bp, "plain") != 0)))
			goto out;
		    *dp = c;
		    for (dp++; isspace (*dp); dp++)
			continue;
		    if (*dp) {
			if ((result = !uprf (dp, "charset")))
			    goto out;
			dp += sizeof("charset") - 1;
			while (isspace (*dp))
			    dp++;
			if (*dp++ != '=')
			    goto invalid;
			while (isspace (*dp))
			    dp++;
			if (*dp == '"') {
			    if ((bp = strchr(++dp, '"')))
				*bp = '\0';
			} else {
			    for (bp = dp; *bp; bp++)
				if (isspace (*bp)) {
				    *bp = '\0';
				    break;
				}
			}
		    } else {
			/* Default character set */
			dp = "US-ASCII";
		    }
		    /* Check the character set */
		    result = !check_charset (dp, strlen (dp));
		} else {
		    if (!(result = (strcasecmp (bp, "text") != 0))) {
			*dp = c;
			bp = dp;
			passno = 2;
			goto again;
		    }
		}
out:
		free (cp);
		if (result) {
		    fclose (fp);
		    return result;
		}
		break;
	    }

	    /*
	     * Check Content-Transfer-Encoding field
	     */
	    if (!strcasecmp (name, ENCODING_FIELD)) {
		cp = add (buf, NULL);
		while (state == FLDPLUS) {
		    state = m_getfld (state, name, buf, sizeof(buf), fp);
		    cp = add (buf, cp);
		}
		for (bp = cp; isspace (*bp); bp++)
		    continue;
		for (dp = bp; istoken (*dp); dp++)
		    continue;
		*dp = '\0';
		result = (strcasecmp (bp, "7bit")
		       && strcasecmp (bp, "8bit")
		       && strcasecmp (bp, "binary"));

		free (cp);
		if (result) {
		    fclose (fp);
		    return result;
		}
		break;
	    }

	    /*
	     * Just skip the rest of this header
	     * field and go to next one.
	     */
	    while (state == FLDPLUS)
		state = m_getfld (state, name, buf, sizeof(buf), fp);
	    break;

	    /*
	     * We've passed the message header,
	     * so message is just text.
	     */
	default:
	    fclose (fp);
	    return 0;
	}
    }
}
